storey/flow.py

```diff
@@ -426,57 +426,80 @@ async def _handle_completed(self, event, response):
-    event.
-    :param max_events: Maximum number of events per emitted batch. Set to None to emit all events in one batch on flow termination.
-    :type max_events: int or None
-    :param timeout_secs: Maximum number of seconds to wait before a batch is emitted.
-    :type timeout_secs: int
-    """
-
-    def __init__(self, max_events: Optional[int], timeout_secs=None, **kwargs):
-        Flow.__init__(self, **kwargs)
+class _Batching(Flow):
+    def __init__(self, max_events: Optional[int] = None, timeout_secs=None, **kwargs):
+        super().__init__(**kwargs)
         self._max_events = max_events
         self._event_count = 0
         self._batch = []
+        self._batch_time = None
         self._timeout_task = None
 
         self._timeout_secs = timeout_secs
         if self._timeout_secs is not None and self._timeout_secs <= 0:
             raise ValueError('Batch timeout cannot be 0 or negative')
 
-    async def sleep_and_emit(self):
+    async def _emit(self, batch, batch_time):
+        raise NotImplementedError
+
+    async def _terminate(self):
+        raise NotImplementedError
+
+    async def _sleep_and_emit(self):
         await asyncio.sleep(self._timeout_secs)
-        await self.emit_batch()
+        await self._emit_batch()
 
     async def _do(self, event):
         if event is _termination_obj:
             if self._timeout_task and not self._timeout_task.cancelled():
                 self._timeout_task.cancel()
-            await self.emit_batch()
-            return await self._do_downstream(_termination_obj)
+            await self._emit_batch()
+            return await self._terminate()
         else:
-            if len(self._batch) == 0 and self._timeout_secs:
-                self._timeout_task = asyncio.get_running_loop().create_task(self.sleep_and_emit())
+            if len(self._batch) == 0:
+                self._batch_time = event.time
+                if self._timeout_secs:
+                    self._timeout_task = asyncio.get_running_loop().create_task(self._sleep_and_emit())
 
             self._event_count = self._event_count + 1
-            self._batch.append(event)
+            self._batch.append(self._get_safe_event_or_body(event))
 
             if self._event_count == self._max_events:
                 if self._timeout_task and not self._timeout_task.cancelled():
                     self._timeout_task.cancel()
-                await self.emit_batch()
+                await self._emit_batch()
 
-    async def emit_batch(self):
+    async def _emit_batch(self):
         if len(self._batch) > 0:
             batch_to_emit = self._batch
+            batch_time = self._batch_time
             self._batch = []
+            self._batch_time = None
             self._event_count = 0
 
-            await self._do_downstream(Event([self._get_safe_event_or_body(event) for event in batch_to_emit], time=batch_to_emit[0].time))
+            await self._emit(batch_to_emit, batch_time)
```

Comment:
why save the batch_time separately, what's wrong with `batch_to_emit[0].time`?

Reply:
The fast that `batch_to_emit[0]` is not an Event unless you pass in `full_event=True`.